In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


In [2]:
ruta_archivo1 = "Datasets/Vehicle Fuel Economy Data.csv"
df = pd.read_csv(ruta_archivo1)

df.head()

/tmp/ipykernel_49827/526747422.py:2: DtypeWarning: Columns (73,75) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(ruta_archivo1)


Year Manufacturer                          Model  barrels08  barrelsA08  \
0  2017          BMW                       M6 Coupe  17.500588         0.0   
1  2017          BMW                   X3 sDrive28i  12.396250         0.0   
2  2016     Infiniti                 Q50S Red Sport  13.523182         0.0   
3  2017         MINI  John Cooper Works Convertible  11.018889         0.0   
4  2017          Kia                      Sedona SX  14.167143         0.0   

   charge240  city08  city08U  cityA08  cityA08U  ...  mfrCode  c240Dscr  \
0        0.0    15.0     14.0      0.0       0.0  ...      BMX       NaN   
1        0.0    21.0     21.0      0.0       0.0  ...      BMX       NaN   
2        0.0    20.0     19.0      0.0       0.0  ...      NSX       NaN   
3        0.0    24.0     23.0      0.0       0.0  ...      BMX       NaN   
4        0.0    18.0     18.0      0.0       0.0  ...      KMX       NaN   

   charge240b  c240bDscr   createdOn  modifiedOn  startStop  phevCity  \
0         0.0        NaN  04-04-2016  09-09-2016          Y       0.0   
1         0.0        NaN  04-04-2016  22-11-2016          Y       0.0   
2         0.0        NaN  04-04-2016  26-09-2016          N       0.0   
3         0.0        NaN  12-04-2016  16-09-2016          Y       0.0   
4         0.0        NaN  12-04-2016  02-12-2016          N       0.0   

   phevHwy  phevComb  
0      0.0       0.0  
1      0.0       0.0  
2      0.0       0.0  
3      0.0       0.0  
4      0.0       0.0  

[5 rows x 82 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46186 entries, 0 to 46185
Data columns (total 82 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Year             46186 non-null  int64  
 1   Manufacturer     46186 non-null  object 
 2   Model            46176 non-null  object 
 3   barrels08        46169 non-null  float64
 4   barrelsA08       46169 non-null  float64
 5   charge240        46169 non-null  float64
 6   city08           46169 non-null  float64
 7   city08U          46169 non-null  float64
 8   cityA08          46169 non-null  float64
 9   cityA08U         46169 non-null  float64
 10  cityCD           46169 non-null  float64
 11  cityE            46169 non-null  float64
 12  cityUF           46169 non-null  float64
 13  co2              46169 non-null  float64
 14  co2A             46169 non-null  float64
 15  co2TailpipeAGpm  46169 non-null  float64
 16  co2TailpipeGpm   46169 non-null  float64
 17  comb08      

In [4]:
df.describe()

Year     barrels08    barrelsA08     charge240        city08  \
count  46186.000000  46169.000000  46169.000000  46169.000000  46169.000000   
mean    2003.733079     15.306719      0.166865      0.114687     19.167493   
std       12.252017      4.343434      0.869974      1.012054     10.448985   
min     1984.000000      0.047081      0.000000      0.000000      6.000000   
25%     1992.000000     12.935217      0.000000      0.000000     15.000000   
50%     2005.000000     14.875500      0.000000      0.000000     18.000000   
75%     2015.000000     17.500588      0.000000      0.000000     21.000000   
max     2024.000000     42.501429     16.000000     19.000000    153.000000   

            city08U       cityA08      cityA08U        cityCD         cityE  \
count  46169.000000  46169.000000  46169.000000  46169.000000  46169.000000   
mean       7.965150      0.843618      0.701921      0.000238      0.701618   
std       14.226484      6.450475      6.316454      0.033236      5.649704   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%       17.000000      0.000000      0.000000      0.000000      0.000000   
max      153.000000    145.000000    145.000000      5.000000    122.000000   

       ...     rangeHwyA         UCity        UCityA      UHighway  \
count  ...  44192.000000  44192.000000  44192.000000  44192.000000   
mean   ...      0.164464     24.554556      1.107372     35.704448   
std    ...      2.477254     15.141158      9.236931     13.782685   
min    ...      0.000000      0.000000      0.000000      0.000000   
25%    ...      0.000000     18.749900      0.000000     28.300000   
50%    ...      0.000000     22.000000      0.000000     34.000000   
75%    ...      0.000000     26.666700      0.000000     40.000000   
max    ...    114.000000    224.800000    207.000000    187.100000   

          UHighwayA  youSaveSpend    charge240b      phevCity       phevHwy  \
count  44192.000000  44192.000000  44170.000000  44170.000000  44170.000000   
mean       0.866718  -5093.925371      0.014150      0.283858      0.286710   
std        5.869861   4258.449576      0.296926      3.845646      3.723701   
min        0.000000 -37250.000000      0.000000      0.000000      0.000000   
25%        0.000000  -7750.000000      0.000000      0.000000      0.000000   
50%        0.000000  -4750.000000      0.000000      0.000000      0.000000   
75%        0.000000  -2250.000000      0.000000      0.000000      0.000000   
max      173.000000   7000.000000      9.000000     97.000000     81.000000   

           phevComb  
count  44170.000000  
mean       0.283903  
std        3.766720  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max       88.000000  

[8 rows x 58 columns]

Al revisar las estadísticas descriptivas, procedemos a eliminar aquellas columnas que cuentan con demasiados ceros y su promedio es muy bajo.

In [5]:
# Lista de columnas a eliminar
columns_to_drop = [
    'barrelsA08', 'charge240', 'cityA08', 'cityA08U', 'cityCD', 'cityE',
    'rangeCity', 'rangeCityA', 'rangeHwy', 'rangeHwyA', 'UCityA', 'UHighwayA',
    'charge240b', 'phevCity', 'phevHwy', 'phevComb'
]

# Eliminar las columnas del DataFrame
df.drop(columns=columns_to_drop, inplace=True)


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46186 entries, 0 to 46185
Data columns (total 66 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Year             46186 non-null  int64  
 1   Manufacturer     46186 non-null  object 
 2   Model            46176 non-null  object 
 3   barrels08        46169 non-null  float64
 4   city08           46169 non-null  float64
 5   city08U          46169 non-null  float64
 6   cityUF           46169 non-null  float64
 7   co2              46169 non-null  float64
 8   co2A             46169 non-null  float64
 9   co2TailpipeAGpm  46169 non-null  float64
 10  co2TailpipeGpm   46169 non-null  float64
 11  comb08           46169 non-null  float64
 12  comb08U          46169 non-null  float64
 13  combA08          46169 non-null  float64
 14  combA08U         46169 non-null  float64
 15  combE            46169 non-null  float64
 16  combinedCD       46169 non-null  float64
 17  combinedUF  

In [9]:
df.rename(columns={
    'Year': 'Año',
    'Manufacturer': 'Fabricante',
    'Model': 'Modelo',
    'barrels08': 'Barriles08',
    'city08': 'Ciudad08',
    'city08U': 'Ciudad08U',
    'cityUF': 'CiudadUF',
    'co2': 'CO2',
    'co2A': 'CO2A',
    'co2TailpipeAGpm': 'CO2ColaAGpm',
    'co2TailpipeGpm': 'CO2ColaGpm',
    'comb08': 'Combinado08',
    'comb08U': 'Combinado08U',
    'combA08': 'CombinadoA08',
    'combA08U': 'CombinadoA08U',
    'combE': 'CombinadoE',
    'combinedCD': 'CombinadoCD',
    'combinedUF': 'CombinadoUF',
    'cylinders': 'Cilindros',
    'displ': 'Desplazamiento',
    'drive': 'Tracción',
    'engId': 'IDMotor',
    'eng_dscr': 'DescripciónMotor',
    'feScore': 'PuntuaciónFE',
    'fuelCost08': 'CostoCombustible08',
    'fuelCostA08': 'CostoCombustibleA08',
    'fuelType': 'TipoCombustible',
    'fuelType1': 'TipoCombustible1',
    'ghgScore': 'PuntuaciónGHG',
    'ghgScoreA': 'PuntuaciónGHGA',
    'highway08': 'Autopista08',
    'highway08U': 'Autopista08U',
    'highwayA08': 'AutopistaA08',
    'highwayA08U': 'AutopistaA08U',
    'VClass': 'ClaseVehículo',
    'highwayCD': 'AutopistaCD',
    'highwayE': 'AutopistaE',
    'highwayUF': 'AutopistaUF',
    'hlv': 'HLV',
    'hpv': 'HPV',
    'id': 'ID',
    'lv2': 'LV2',
    'lv4': 'LV4',
    'mpgData': 'DatosMPG',
    'phevBlended': 'HíbridoEnchufable',
    'pv2': 'PV2',
    'pv4': 'PV4',
    'range': 'Rango',
    'trany': 'Transmisión',
    'UCity': 'UCiudad',
    'UHighway': 'UAutopista',
    'youSaveSpend': 'AhorroGasto',
    'guzzler': 'Guzzler',
    'trans_dscr': 'DescripciónTrans',
    'tCharger': 'TurboCargador',
    'sCharger': 'Supercargador',
    'atvType': 'TipoATV',
    'fuelType2': 'TipoCombustible2',
    'rangeA': 'RangoA',
    'evMotor': 'MotorEV',
    'mfrCode': 'CódigoFabricante',
    'c240Dscr': 'DescripciónC240',
    'c240bDscr': 'DescripciónC240b',
    'createdOn': 'CreadoEn',
    'modifiedOn': 'ModificadoEn',
    'startStop': 'ArranqueParada'
}, inplace=True)


In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46186 entries, 0 to 46185
Data columns (total 66 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Año                  46186 non-null  int64  
 1   Fabricante           46186 non-null  object 
 2   Modelo               46176 non-null  object 
 3   Barriles08           46169 non-null  float64
 4   Ciudad08             46169 non-null  float64
 5   Ciudad08U            46169 non-null  float64
 6   CiudadUF             46169 non-null  float64
 7   CO2                  46169 non-null  float64
 8   CO2A                 46169 non-null  float64
 9   CO2ColaAGpm          46169 non-null  float64
 10  CO2ColaGpm           46169 non-null  float64
 11  Combinado08          46169 non-null  float64
 12  Combinado08U         46169 non-null  float64
 13  CombinadoA08         46169 non-null  float64
 14  CombinadoA08U        46169 non-null  float64
 15  CombinadoE           46169 non-null 

Nulos

In [19]:
# Contar los valores nulos por columna
nulos_por_columna = df.isnull().sum()


In [20]:
# Configurar la opción para mostrar todas las columnas
pd.set_option('display.max_rows', None)

# Mostrar los valores nulos por columna
print(nulos_por_columna)


Año                        0
Fabricante                 0
Modelo                    10
Barriles08                17
Ciudad08                  17
Ciudad08U                 17
CiudadUF                  17
CO2                       17
CO2A                      17
CO2ColaAGpm               17
CO2ColaGpm                17
Combinado08               17
Combinado08U              17
CombinadoA08              17
CombinadoA08U             17
CombinadoE                17
CombinadoCD               17
CombinadoUF               17
Cilindros                523
Desplazamiento           521
Tracción                1203
IDMotor                   17
DescripciónMotor       17094
PuntuaciónFE            1233
CostoCombustible08      1233
CostoCombustibleA08     1233
TipoCombustible         1233
TipoCombustible1        1233
PuntuaciónGHG           1233
PuntuaciónGHGA          1233
Autopista08             1233
Autopista08U            1233
AutopistaA08            1233
AutopistaA08U           1233
ClaseVehículo 

In [21]:
# Eliminar las columnas del DataFrame
df.drop(columns=['DescripciónC240', 'DescripciónC240b'], inplace=True)


Estadísticas descriptivas:

- Año: Los vehículos en el dataset abarcan desde el año 1984 hasta el 2024, con un promedio de 2003.
- Barriles08: El consumo anual promedio de barriles de petróleo por vehículo es de 15.31, con una desviación estándar de 4.34.
- Ciudad08: El promedio de millas por galón (MPG) en ciudad es de 19.17, con una desviación estándar de 10.45.
- Ciudad08U: El promedio de MPG en ciudad sin redondear es de 7.97, con una desviación estándar de 14.23.
- CiudadUF: Todos los valores son cero, lo que indica que esta columna puede no ser relevante o estar mal registrada.
- CO2: El promedio de emisiones de CO2 en gramos por milla es de 120.10, con una desviación estándar de 194.61.
- CO2A: El promedio de emisiones alternativas de CO2 en gramos por milla es de 5.85, con una desviación estándar de 57.08.
- CO2ColaAGpm: El promedio de emisiones alternativas de CO2 por galón de combustible es de 16.54, con una desviación estándar de 90.67.
- CO2ColaGpm: El promedio de emisiones de CO2 por galón de combustible es de 457.01, con una desviación estándar de 128.68.
- Combinado08: El promedio de MPG combinado (ciudad + carretera) es de 21.38, con una desviación estándar de 9.89.
- HPV: El promedio de potencia híbrida en vehículos es de 9.32, con una desviación estándar de 26.98.
- ID: El ID promedio es de 23602.57, con una desviación estándar de 13483.47.
- LV2: El promedio de nivel 2 es de 1.75, con una desviación estándar de 4.28.
- LV4: El promedio de nivel 4 es de 6.03, con una desviación estándar de 9.52.
- PV2: El promedio de potencia 2 es de 13.22, con una desviación estándar de 30.74.
- PV4: El promedio de potencia 4 es de 33.67, con una desviación estándar de 46.16.
- Rango: El promedio de rango es de 2.57, con una desviación estándar de 26.04.
- UCiudad: El promedio de MPG en ciudad sin redondear y ajustado para reflejar el uso real es de 24.55, con una desviación estándar de 15.14.
- UAutopista: El promedio de MPG en carretera sin redondear y ajustado para reflejar el uso real es de 35.70, con una desviación estándar de 13.78.
- AhorroGasto: El promedio de ahorro o gasto en comparación con el vehículo promedio es de -5093.93, con una desviación estándar de 4258.45.

Algunas consideraciones generales sobre las estadísticas descriptivas:

- Variabilidad en la Eficiencia de Combustible: La diferencia entre el mínimo y el máximo en las columnas de eficiencia de combustible (Ciudad08, Combinado08 y UAutopista) sugiere una amplia variabilidad en la eficiencia de combustible de los vehículos en el dataset. Esto podría indicar la presencia de varios tipos de vehículos, desde aquellos que son muy eficientes hasta los que consumen mucho combustible.

- Tendencia hacia Vehículos más Eficientes: Dado que el año promedio de los vehículos es 2003, pero el rango se extiende hasta 2024, es posible que haya una tendencia hacia una mayor eficiencia de combustible y menores emisiones de CO2 en los modelos más recientes.

- Ahorro o Gasto en Combustible: La columna AhorroGasto muestra una amplia gama de valores, desde un ahorro significativo hasta un gasto considerable en comparación con el vehículo promedio. Esto refleja la diversidad en la eficiencia de combustible de los vehículos y puede ser un factor importante para los consumidores al elegir un vehículo.

In [11]:
df.describe()

Año    Barriles08      Ciudad08     Ciudad08U  CiudadUF  \
count  46186.000000  46169.000000  46169.000000  46169.000000   46169.0   
mean    2003.733079     15.306719     19.167493      7.965150       0.0   
std       12.252017      4.343434     10.448985     14.226484       0.0   
min     1984.000000      0.047081      6.000000      0.000000       0.0   
25%     1992.000000     12.935217     15.000000      0.000000       0.0   
50%     2005.000000     14.875500     18.000000      0.000000       0.0   
75%     2015.000000     17.500588     21.000000     17.000000       0.0   
max     2024.000000     42.501429    153.000000    153.000000       0.0   

                CO2          CO2A   CO2ColaAGpm    CO2ColaGpm   Combinado08  \
count  46169.000000  46169.000000  46169.000000  46169.000000  46169.000000   
mean     120.104811      5.853950     16.535381    457.010711     21.376985   
std      194.612428     57.077069     90.666005    128.683468      9.894844   
min       -1.000000     -1.000000      0.000000      0.000000      7.000000   
25%       -1.000000     -1.000000      0.000000    380.000000     17.000000   
50%       -1.000000     -1.000000      0.000000    444.350000     20.000000   
75%      311.000000     -1.000000      0.000000    522.764706     23.000000   
max      979.000000    713.000000    713.000000   1269.571429    142.000000   

       ...           HPV            ID           LV2           LV4  \
count  ...  44192.000000  44192.000000  44192.000000  44192.000000   
mean   ...      9.324108  23602.573588      1.746040      6.031092   
std    ...     26.977040  13483.466970      4.280382      9.518442   
min    ...      0.000000      1.000000      0.000000      0.000000   
25%    ...      0.000000  11960.750000      0.000000      0.000000   
50%    ...      0.000000  23637.500000      0.000000      0.000000   
75%    ...      0.000000  35441.250000      0.000000     13.000000   
max    ...    195.000000  46624.000000     41.000000     55.000000   

                PV2           PV4         Rango       UCiudad    UAutopista  \
count  44192.000000  44192.000000  44192.000000  44192.000000  44192.000000   
mean      13.221669     33.674556      2.566573     24.554556     35.704448   
std       30.736695     46.162066     26.035835     15.141158     13.782685   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000     18.749900     28.300000   
50%        0.000000      0.000000      0.000000     22.000000     34.000000   
75%        0.000000     91.000000      0.000000     26.666700     40.000000   
max      194.000000    192.000000    520.000000    224.800000    187.100000   

        AhorroGasto  
count  44192.000000  
mean   -5093.925371  
std     4258.449576  
min   -37250.000000  
25%    -7750.000000  
50%    -4750.000000  
75%    -2250.000000  
max     7000.000000  

[8 rows x 42 columns]

Cruce de datos

Los 10 años con màs contaminaciòn ordenados de mayor a menor

In [12]:
# Agrupar por año y calcular el promedio de emisiones de CO2
co2_by_year = df.groupby('Año')['CO2'].mean()

# Ordenar los resultados de mayor a menor
co2_by_year_sorted = co2_by_year.sort_values(ascending=False)

# Mostrar los años con más contaminación
print(co2_by_year_sorted.head(10))


Año
2013    420.664407
2014    414.313677
2015    404.960938
2017    392.502707
2016    392.299525
2018    390.081663
2019    388.092399
2021    388.017969
2020    387.362620
2022    378.605323
Name: CO2, dtype: float64


Los 10 años con más ahorro de combustible

In [13]:

# Agrupar por año y calcular el promedio de ahorro de combustible
ahorro_by_year = df.groupby('Año')['AhorroGasto'].mean()

# Ordenar los resultados de mayor a menor ahorro
ahorro_by_year_sorted = ahorro_by_year.sort_values(ascending=False)

# Mostrar los 10 años con más ahorro de combustible
print(ahorro_by_year_sorted.head(10))


Año
2023   -3405.877681
2024   -3564.338235
2022   -3574.067022
2016   -3631.568471
2017   -3636.557887
2019   -3645.444361
2018   -3690.476190
2020   -3779.400000
2021   -3810.250391
2015   -3922.543582
Name: AhorroGasto, dtype: float64


Los 10 autos más contaminantes

In [14]:
# Ordenar el DataFrame por emisiones de CO2 de mayor a menor
df_sorted = df.sort_values(by='CO2ColaGpm', ascending=False)

# Seleccionar los 10 modelos más contaminantes
top10_contaminantes = df_sorted[['Modelo', 'CO2ColaGpm']].head(10)

# Mostrar los resultados
print(top10_contaminantes)


                          Modelo   CO2ColaGpm
45351                   Countach  1269.571429
914                     Countach  1269.571429
14777                   Countach  1269.571429
11445                   Countach  1269.571429
26562                   Countach  1269.571429
3385   Silver Spirit/Silver Spur  1110.875000
4036     Corniche II/Continental  1110.875000
6357                 Ferrari F50  1110.875000
45184      Silver Spur Limousine  1110.875000
30676                   Camargue  1110.875000


Los 10 modelos menos contaminantes

In [15]:
# Ordenar el DataFrame por emisiones de CO2 de menor a mayor
df_sorted = df.sort_values(by='CO2ColaGpm', ascending=True)

# Seleccionar los 10 autos menos contaminantes
bottom10_contaminantes = df_sorted[['Modelo', 'CO2ColaGpm']].head(10)

# Mostrar los resultados
print(bottom10_contaminantes)


                                        Modelo  CO2ColaGpm
28969        Model S Performance (19in Wheels)         0.0
29259             Taycan Turbo S Cross Turismo         0.0
19551        fortwo electric drive convertible         0.0
20953                       Model 3 Long Range         0.0
10927       i4 M50 Gran Coupe (19 inch wheels)         0.0
20405                        Model S AWD - 70D         0.0
22256  i4 eDrive40 Gran Coupe (18 inch wheels)         0.0
31281        fortwo electric drive convertible         0.0
20404          Model S (70 kW-hr battery pack)         0.0
38289                                   e-tron         0.0
